File: src/main/java/com/github/wolray/line/io/ValuesJoiner.java
package com.github.wolray.line.io;

import com.github.wolray.seq.ArraySeq;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.function.Function;
import java.util.function.Predicate;

import static com.github.wolray.line.io.TypeValues.invoke;

/**
 * @author wolray
 */
public class ValuesJoiner<T> {
    private final TypeValues<T> typeValues;
    private final ArraySeq<TypeValues.Attr> attrs;

    public ValuesJoiner(TypeValues<T> typeValues) {
        this.typeValues = typeValues;
        attrs = typeValues.toAttrs();
        initFormatters();
    }

    private void initFormatters() {
        Function<Object, String> toString = Object::toString;
        attrs.supply(a -> a.formatter = toString);
        TypeValues.processSimpleMethods(typeValues.type, this::processMethod);
    }

    void processMethod(TypeValues.SimpleMethod simpleMethod) {
        Method method = simpleMethod.method;
        Class<?> paraType = simpleMethod.paraType;
        if (paraType != String.class && simpleMethod.returnType == String.class) {
            method.setAccessible(true);
            Function<Object, String> function = s -> (String)invoke(method, s);
            Fields fields = method.getAnnotation(Fields.class);
            Predicate<Field> predicate = FieldSelector.toPredicate(fields);
            attrs
                .filter(a -> predicate.test(a.field) && a.field.getType() == paraType)
                .supply(a -> a.formatter = function);
        }
    }

    String join(String sep) {
        return join(sep, a -> a.field.getName());
    }

    String join(String sep, Function<TypeValues.Attr, String> function) {
        return attrs.join(sep, function);
    }

    public Function<T, String> toFormatter(String sep) {
        return t -> join(sep, a -> a.format(a.get(t)));
    }
}

File: src/main/java/com/github/wolray/line/io/ValuesJoiner.kt
package com.github.wolray.line.io

import com.github.wolray.line.io.DataMapper.Companion.toTest
import com.github.wolray.line.io.MethodScope.annotation
import com.github.wolray.line.io.MethodScope.asMapper
import com.github.wolray.line.io.TypeScope.isType
import com.github.wolray.line.io.TypeValues.SimpleMethod
import com.github.wolray.seq.ArraySeq
import java.lang.reflect.Field
import java.util.function.Function

/**
 * @author wolray
 */
class ValuesJoiner<T>(typeValues: TypeValues<T>) {
    val attrs: ArraySeq<Attr> = typeValues.map { it -> Attr(it, mapper) }.toList()

    init {
        TypeValues.processSimpleMethods(typeValues.type, ::processMethod)
    }

    private fun processMethod(simpleMethod: SimpleMethod) {
        val method = simpleMethod.method
        val paraType = simpleMethod.paraType
        if (paraType.isType<String>().not() && simpleMethod.returnType.isType<String>()) {
            val mapper = method.asMapper<Any?, String>("")
            val test = method.annotation<Fields>().toTest()
            attrs
                .filter { a -> test(a.field) && a.field.type == paraType }
                .supply { a -> a.mapper = mapper }
        }
    }

    fun toFormatter(sep: String): Function<T, String> = Function { t ->
        join(sep) { it.mapper(it.field[t]) }
    }

    fun joinFields(sep: String) = join(sep) { it.field.name }

    fun join(sep: String, mapper: (Attr) -> String) = attrs.joinToString(sep, transform = mapper)

    class Attr(val field: Field, var mapper: (Any?) -> String)

    companion object {
        val mapper = { o: Any? -> o?.toString().orEmpty() }
    }
}